kong/resty/dns_client/README.md

```diff
@@ -0,0 +1,117 @@
+
+The module is currently Kong only, and builds on top of the `lua-resty-dns` and the kong's `lua-resty-mlcache` library.
+
+Table of Contents
+=================
+
+* [Name](#name)
+* [APIs](#apis)
+    * [new](#new)
+    * [resolve](#resolve)
+
+# APIs
+
+The following APIs are for internal development use only within Kong. In the current version, the new DNS library still needs to be compatible with the original DNS library. Therefore, the functions listed below cannot be directly invoked. For example, the `_M:resolve` function in the following APIs will be replaced to ensure compatibility with the previous DNS library API interface specifications `_M.resolve`.
+
+## new
+
+**syntax:** *c, err = dns_client.new(opts)*  
+**context:** any
+
+**Functionality:**
+
+Creates a dns client object. Returns nil and a message string on error.
+
+Perform a series of initialization operations:
+
+* parse `host` file
+* parse `resolv.conf` file (used by the underlying `lua-resty-dns` library)
+* initialize multiple TTL options
+* create a mlcache object and initialize it
+
+**Input paramenters:**
+
+`@opts` It accepts a options table argument. The following options are supported:
+
+* TTL options
+  * `valid_ttl`
+    * same to the option `dns_valid_ttl` in kong.conf
+  * `stale_ttl`
+    * same to the option `dns_stale_ttl` in kong.conf
+  * `empty_ttl`
+    * same to the option `dns_not_found_ttl` in kong.conf
+  * `bad_ttl`
+    * same to the option `dns_error_ttl` in kong.conf
+* `hosts` (default: `/etc/hosts`)
+  * the path of `hosts` file
+* `resolv_conf` (default: `/etc/resolv.conf`)
+  * the path of `resolv.conf` file, it will be parsed and passed into the underlying `lua-resty-dns` library.
+* `order` (default: `{ "LAST", "SRV", "A", "AAAA", "CNAME" }`)
+  * the order in which to resolve different record types, it's similar to the option `dns_order` in kong.conf.
+  * The `LAST` type means the type of the last successful lookup (for the specified name).
+* `enable_ipv6` (default: `ture`)
+  * whether to support IPv6 servers when when getting nameservers from `resolv.conf`
+* options for the underlying `lua-resty-dns` library 
+  * `retrans` (default: `5`)
+    * the total number of times of retransmitting the DNS request when receiving a DNS response times out according to the timeout setting. When trying to retransmit the query, the next nameserver according to the round-robin algorithm will be picked up.
+    * If not given, it is taken from `resolv.conf` option `options attempts:<value>`.
+  * `timeout` (default: `2000`)
+    * the time in milliseconds for waiting for the response for a single attempt of request transmission
+    * If not given, it is taken from `resolv.conf` option `options timeout:<value>`. But note that its unit in `resolv.conf` is second.
+  * `no_random` (default: `true`)
+    * a boolean flag controls whether to randomly pick the nameserver to query first, if `true` will always start with the first nameserver listed. 
```

Comment:
```suggestion
    * a boolean flag controls whether to randomly pick the nameserver to query first, if `true` it will always start with the first nameserver listed. 
```